[24.0] Fix tool panel loading delay

ahmedhamidawan opened the pull request:
Improved client performance by:
1. Keeping the tool search worker in the `toolStore`
2. Removing unnecessary computed variables that were causing long load times and instead refactoring them to be `shallowRef` or sets

Fixes https://github.com/galaxyproject/galaxy/issues/17825

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

martenson commented on client/src/components/Panels/ToolBox.vue:
```diff
@@ -89,33 +89,24 @@ const dataManagerSection = computed(() => {
-});
+const localToolsById =
+    toolStore.toolsById && Object.keys(toolStore.toolsById).length > 0
+        ? getValidToolsInCurrentView(toolStore.toolsById, props.workflow, !props.workflow ? SECTION_IDS_TO_EXCLUDE : [])
```
this is imho very hard to read

ahmedhamidawan replied:
I hope the readability is better now

ahmedhamidawan committed:
Improve panel speed

Co-authored-by: mvdbeek <m.vandenbeek@gmail.com>

mvdbeek committed:
Optimize getValidToolsInCurrentView

mvdbeek committed:
Optimize getValidToolsInEachSection

mvdbeek committed:
Make toolsById a shallowRef

A given tool never changes attributes (currently), so no need to track
reactivity for the whole object.

ahmedhamidawan committed:
keep tool search worker in `toolStore` to prevent multiple bundle fetches

mvdbeek commented on client/src/stores/toolStore.ts:
```diff
@@ -75,12 +75,19 @@ export interface PanelView {
+    // TODO: Where do we terminate the worker?
+    // onScopeDispose(() => {
+    //     searchWorker.value?.terminate();
+    //     searchWorker.value = undefined;
```
How important is it to terminate the worker ? It does seem like the callback holds on to `currentFavorites.value.tools` (which makes we wonder, does the reactivity there work cross thread?), but if we only ever create a single worker that seems fine ? Do you know if the callback keeps a reference to the ToolSearch Component ?

Maybe you could remove just the callback when you unmount the tool search component ?

ahmedhamidawan replied:
We definitely only ever create the one worker here, because we only create it if it is undefined in the store.

I'm trying to investigate where you're getting the value for the "callback" though? Do you mean the onmessage callback?

ahmedhamidawan replied:
So, upon investigation, I can see that the onmessage handler retains its value outside the scope of the `ToolSearch` component. Does it keep a "reference" to the refs inside of it outside the component, I can't seem to be able to inspect that but to be safe: I'll do what you recommended here and just remove the callback `onUnmounted` since there is no harm in that.

ahmedhamidawan committed:
set tool search worker to `null` to remove references while unmounted
